pyiron_continuum/schroedinger/schroedinger.py

```diff
@@ -0,0 +1,278 @@
+    )
+
+
+# TODO: Convert to pyiron units
+HBAR = 1  # EV_TO_U_ANGSTROMSQ_PER_SSQ * physical_constants['Planck constant in eV/Hz'][0] / (2 * np.pi)
+M_EL = 1  # physical_constants['electron mass in u'][0]
+KB = physical_constants['Boltzmann constant in eV/K'][0]
+
+__author__ = "Liam Huber"
+__copyright__ = (
+    "Copyright 2021, Max-Planck-Institut für Eisenforschung GmbH - "
+    "Computational Materials Design (CM) Department"
+)
+__version__ = "0.1"
+__maintainer__ = "Liam Huber"
+__email__ = "huber@mpie.de"
+__status__ = "development"
+__date__ = "Sep 8, 2021"
+
+
+class _TISEInput(DataContainer):
+    def __init__(self, init=None, table_name='input', lazy=False):
+        super().__init__(init=init, table_name=table_name, lazy=lazy)
+        self._mesh = None
+        self._potential = None
+        self.n_states = 1
+        self.mass = M_EL
+
+    @property
+    def mesh(self) -> RectMesh:
+        return self._mesh
+
+    @mesh.setter
+    def mesh(self, new_mesh: RectMesh):
+        if not isinstance(new_mesh, RectMesh):
+            raise TypeError(f'Meshes must be of type {RectMesh} but got type {type(new_mesh)}')
+        new_mesh.simplify_1d = False
+        self._mesh = new_mesh
+
+    @property
+    def potential(self):
+        return self._potential
+
+    @potential.setter
+    def potential(self, new_potential):
+        self._potential = new_potential
+
+
+class _TISEOutput(DataContainer):
+    def __init__(self, init=None, table_name='input', lazy=False):
+        super().__init__(init=init, table_name=table_name, lazy=lazy)
+        self.energy = None
+        self.psi = None
+        self._rho = None
+
+    @property
+    def _space_axes(self):
+        return tuple(n for n in range(1, len(self.psi.shape)))
+
+    def _broadcast_vector_to_space(self, vector):
+        return vector.reshape(-1, *(1,)*len(self._space_axes))
+
+    def _normalize_per_state(self, states, ord=2):
+        return states / self._broadcast_vector_to_space(np.linalg.norm(states, axis=self._space_axes, ord=ord))
+
+    def _weight_states(self, states, weights):
+        return states * self._broadcast_vector_to_space(weights)
+
+    @property
+    def rho(self):
+        if self._rho is None and self.psi is not None:
+            self._rho = self._normalize_per_state(self.psi ** 2, ord=1)
+        return self._rho
+
+    def get_boltzmann_occupation(self, temperature):
+        if self.energy is not None:
+            return np.exp(-self.energy / (KB * temperature))
+
+    def get_boltzmann_psi(self, temperature):
+        if self.psi is not None:
+            weighted_psi = self._weight_states(self.psi, self.get_boltzmann_occupation(temperature))
+            psi_tot = np.sum(weighted_psi, axis=0)
+            return psi_tot / np.linalg.norm(psi_tot)
+
+    def get_boltzmann_rho(self, temperature):
+        if self.psi is not None:
+            weighted_rho = self._weight_states(self.rho, self.get_boltzmann_occupation(temperature)**2)
```

Comment:
I see here that you have squared the Boltzmann weights. Doesn't doing this give the same result as using the hidden method below (`_boltzmann_rho_from_psi`)?

Reply:
Invoke both on the same system and check ;)
(But seriously, I intend to do that in an integration test to make sure it stays true forever.)

Reply:
Okay, yes, does NOT give the same result. However, is there a physical reason for doing this? It gives me a quantum solution at high temperatures that does not agree with the classical solution, and it bothers me. (I make the same face as the cat in my picture). NOT squaring the B weight gives me an agreement.

Reply:
Aha, I got it. Give me a minute to write it up, but we are both right 😉

Reply:
I just realized that you might be looking at Griffiths. I remembered reading a couple of days ago that the weight factor (which I think he calls `c_n`) should be `abs(c_n**2)`, which would then mean that `c_n` should be L2-normed. I don't have the text book at hand to confirm, but I now see why you had the square in the Boltzmann weights.

Reply:
Yes, I extended the notebook example to show exactly what you mean about $\rho(T) = \sum_i w_i \rho_i$ converging nicely to the classical result -- so unless we're calculating the classical result wrong (e.g. there's a missing factor of two because of energy partition or something) -- then it must ultimately come out that way.

What's driving me crazy is that the boltzmann weight should be the probability to be in *state* *i*, and *state i* is $\psi_i$, so that's what should receive the scale, not $\rho_i$. However, (a) I haven't been able to build a properly symmetric wavefunction to get any sort of reasonable wavefunction out, and (b) I guess it should all just come out to the known answer of linearly weighting the densities anyway since we already know that works...

Reply:
Ugh, ok, I think I see it. The problem is that I just don't know anything about quantum mechanics. In $\Psi = \sum_i c_i \psi_i$, of course $c_i$ is not the probability of being in state $i$ -- how could it be? $c_i$ is allowed to be imaginary! Rather $|c_i|^2 = P(i)$. So of course we shouldn't use $c_i \propto \exp(-E_i / k_B T$ because the *probability* is proportional to this! $P(i) \propto \exp(-E_i / k_B T) \propto |c_i|^2 \rightarrow c_i \propto \exp(-0.5 * E_i / k_B T$. And then of course $rho_i \propto \exp(-E_i / k_B T)$, not $\exp(-2E_i / k_B T)$ like I was originally insisting.

Griffiths eq. 2.39, and the text at the bottom of p. 36 (2nd ed). I haven't felt this dumb in a while.

I would feel better yet if I could cook up a symmetric $\Psi(T)$ to go along with all this, but the argument $|c_i|^2$ gives probability and probability is proportional to  $\exp(-E_i / k_B T)$ so $c_i$ must be proportional to  $\exp(-0.5E_i / k_B T)$ s.t. $rho_i$ is proportional to  $\exp(-2E_i / k_B T)$ is perfectly satisfactory to me.

Reply:
> Griffiths eq. 2.39

Okay, I am glad that the math makes sense! I follow your justification. I looked at the entire bunch of equations in that chapter for far too long on my digital version of the text. I am now quite done with looking at equations for the day.

I shall move on to cleaning up my notebooks and generating some plots now. I will make a PR to this branch in the weekend, since I need to get some BLaSA work done tomorrow. Till then, I'll continue using my notebook class (which really gives the exact same results as the Schroedinger class).